File: src/app/services/metric.service.ts
import { Injectable } from '@angular/core';
import { LocationService } from './location.service';
import { HeartRateSensorService } from './heart-rate-sensor.service';
import { ServiceFactory } from './ServiceFactory';
import { Metric, MetricType } from '../model/Metric';
import { AltitudeMetric } from '../model/AltitudeMetric';
import { VerticalSpeedMetric } from '../model/VerticalSpeedMetric';
import { TemperatureMetric } from '../model/TemperatureMetric';
import { SpeedMetric } from '../model/SpeedMetric';
import { PowerMetric } from '../model/PowerMetric';
import { LongitudeMetric } from '../model/LongitudeMetric';
import { LatitudeMetric } from '../model/LatitudeMetric';
import { HeartRateMetric } from '../model/HeartRateMetric';
import { GradeMetric } from '../model/GradeMetric';
import { DistanceMetric } from '../model/DistanceMetric';
import { CadenceMetric } from '../model/CadenceMetric';
import { IMetricService } from './IMetricService';
import { PowerMeterService } from './power-meter.service';
import { PowerBalenceMetric } from '../model/PowerBalenceMetric';
import { SpeedSensorService } from './speed-sensor.service';
import { WheelRotationsMetric } from '../model/WheelRotationsMetric';
import { BearingMetric } from '../model/BearingMetric';
import { LapCounterMetric } from '../model/LapCounterMetric';

@Injectable({
  providedIn: 'root'
})
export class MetricService implements IMetricService {
  private locationService: LocationService;
  private heartRateSensorService: HeartRateSensorService;
  private powerMeterService: PowerMeterService;
  private speedSensorService: SpeedSensorService;
  private metrics: Metric[] = [];
  private running: boolean = false;
  private lapStartTimes: Date[] = []; // Store lap timestamps

  constructor() {
    this.locationService = ServiceFactory.getLocationService();
    this.heartRateSensorService = ServiceFactory.getHeartRateSensorService();
    this.powerMeterService = ServiceFactory.getPowerMeterService();
    this.speedSensorService = ServiceFactory.getSpeedSensorService();

    this.metrics.push(
      new AltitudeMetric(this),
      new VerticalSpeedMetric(this),
      new TemperatureMetric(this),
      new SpeedMetric(this),
      new PowerMetric(this),
      new LongitudeMetric(this),
      new LatitudeMetric(this),
      new HeartRateMetric(this),
      new GradeMetric(this),
      new DistanceMetric(this),
      new CadenceMetric(this),
      new PowerBalenceMetric(this),
      new WheelRotationsMetric(this),
      new BearingMetric(this),
      new LapCounterMetric(this)
    );
  }

  newLap(): void {
    this.lapStartTimes.push(new Date());
    for (let i = 0; i < this.metrics.length; i++) {
      this.metrics[i].newLap();
    }
  }

  getNumberOfLaps(): number {
    return this.lapStartTimes.length;
  }

  getLapStartTime(lapIndex: number): Date | null {
    if (lapIndex < 0 || lapIndex >= this.lapStartTimes.length) {
      throw new Error(`Lap index ${lapIndex} is out of bounds. Total laps: ${this.lapStartTimes.length}`);
    }
    return this.lapStartTimes[lapIndex];
  }

  displayLapDuration(lapIndex: number): string {
    if (lapIndex < 0 || lapIndex >= this.lapStartTimes.length) {
      throw new Error(`Lap index ${lapIndex} is out of bounds. Total laps: ${this.lapStartTimes.length}`);
    }
    const startTime = this.lapStartTimes[lapIndex];
    const endTime = lapIndex < this.lapStartTimes.length - 1 ? this.lapStartTimes[lapIndex + 1] : new Date();
    const duration = endTime.getTime() - startTime.getTime();
    const seconds = Math.floor((duration / 1000) % 60);
    const minutes = Math.floor((duration / (1000 * 60)) % 60);
    const hours = Math.floor((duration / (1000 * 60 * 60)) % 24);
    if (hours > 0) {
      return `${hours.toString().padStart(2, '0')}:${minutes.toString().padStart(2, '0')}:${seconds.toString().padStart(2, '0')}`;
    }
    return `${minutes.toString().padStart(2, '0')}:${seconds.toString().padStart(2, '0')}`;
  }

  startLogging(): void {
    if (this.running) {
      console.warn('Logging is already running');
      return;
    }

    this.running = true;
    for (let i = 0; i < this.metrics.length; i++) {
      this.metrics[i].startLogging();
    }
    if (this.lapStartTimes.length === 0) {
      this.lapStartTimes.push(new Date()); // Initialize first lap start time
    }
  }

  stopLogging(): void {
    if (!this.running) {
      console.warn('Logging is not running');
      return;
    }
    this.running = false;
    for (let i = 0; i < this.metrics.length; i++) {
      this.metrics[i].stopLogging();
    }
  }

  isRunning(): boolean {
    return this.running;
  }

  getLocationService(): LocationService {
    return this.locationService;
  }

  getHeartRateSensorService(): HeartRateSensorService {
    return this.heartRateSensorService;
  }

  getPowerMeterService(): PowerMeterService {
    return this.powerMeterService;
  }

  getSpeedSensorService(): SpeedSensorService {
    return this.speedSensorService;
  }

  getByName(name: string): Metric | null {
    for (let i = 0; i < this.metrics.length; i++) {
      if (this.metrics[i].getName() === name) {
        return this.metrics[i];
      }
    }
    return null;
  }

  getNames(): string[] {
    const names: string[] = [];
    for (let i = 0; i < this.metrics.length; i++) {
      names.push(this.metrics[i].getName());
    }
    return names;
  }

  getByMetricType(metricType: MetricType): Metric {
    for (let i = 0; i < this.metrics.length; i++) {
      if (this.metrics[i].getType() === metricType) {
        return this.metrics[i];
      }
    }
    throw new Error(`Metric with type ${metricType} not found`);
  }

}

File: android/app/src/main/java/com/openbikecomputer/service/MetricService.java
package com.openbikecomputer.service;

import com.openbikecomputer.model.AltitudeMetric;
import com.openbikecomputer.model.BearingMetric;
import com.openbikecomputer.model.CadenceMetric;
import com.openbikecomputer.model.CurrentTimeMetric;
import com.openbikecomputer.model.DistanceMetric;
import com.openbikecomputer.model.ElapsedTimeMetric;
import com.openbikecomputer.model.GradeMetric;
import com.openbikecomputer.model.HeartRateMetric;
import com.openbikecomputer.model.LapCounterMetric;
import com.openbikecomputer.model.LatitudeMetric;
import com.openbikecomputer.model.LongitudeMetric;
import com.openbikecomputer.model.Metric;
import com.openbikecomputer.model.MetricType;
import com.openbikecomputer.model.PowerBalenceMetric;
import com.openbikecomputer.model.PowerMetric;
import com.openbikecomputer.model.SpeedMetric;
import com.openbikecomputer.model.TemperatureMetric;
import com.openbikecomputer.model.VerticalSpeedMetric;
import com.openbikecomputer.model.WheelRotationsMetric;
import com.openbikecomputer.util.UnitToString;

import java.util.ArrayList;
import java.util.List;

/**
 * Owns the full set of {@link Metric}s and the lap timeline.
 * Ported from {@code metric.service.ts}.
 */
public class MetricService implements IMetricService {

    private final ILocationService locationService;
    private final IHeartRateSensorService heartRateSensorService;
    private final IPowerMeterService powerMeterService;
    private final ISpeedSensorService speedSensorService;

    private final List<Metric> metrics = new ArrayList<>();
    private boolean running = false;
    /** Lap start times as epoch milliseconds. */
    private final List<Long> lapStartTimes = new ArrayList<>();

    public MetricService(ILocationService locationService,
                         IHeartRateSensorService heartRateSensorService,
                         IPowerMeterService powerMeterService,
                         ISpeedSensorService speedSensorService) {
        this.locationService = locationService;
        this.heartRateSensorService = heartRateSensorService;
        this.powerMeterService = powerMeterService;
        this.speedSensorService = speedSensorService;

        metrics.add(new AltitudeMetric(this));
        metrics.add(new VerticalSpeedMetric(this));
        metrics.add(new TemperatureMetric(this));
        metrics.add(new SpeedMetric(this));
        metrics.add(new PowerMetric(this));
        metrics.add(new LongitudeMetric(this));
        metrics.add(new LatitudeMetric(this));
        metrics.add(new HeartRateMetric(this));
        metrics.add(new GradeMetric(this));
        metrics.add(new DistanceMetric(this));
        metrics.add(new CadenceMetric(this));
        metrics.add(new PowerBalenceMetric(this));
        metrics.add(new WheelRotationsMetric(this));
        metrics.add(new BearingMetric(this));
        metrics.add(new LapCounterMetric(this));
        metrics.add(new CurrentTimeMetric(this));
        metrics.add(new ElapsedTimeMetric(this));
    }

    public void newLap() {
        lapStartTimes.add(System.currentTimeMillis());
        for (Metric metric : metrics) {
            metric.newLap();
        }
    }

    public int getNumberOfLaps() {
        return lapStartTimes.size();
    }

    public Long getLapStartTime(int lapIndex) {
        if (lapIndex < 0 || lapIndex >= lapStartTimes.size()) {
            throw new IllegalArgumentException("Lap index " + lapIndex
                    + " is out of bounds. Total laps: " + lapStartTimes.size());
        }
        return lapStartTimes.get(lapIndex);
    }

    public String displayLapDuration(int lapIndex) {
        if (lapIndex < 0 || lapIndex >= lapStartTimes.size()) {
            throw new IllegalArgumentException("Lap index " + lapIndex
                    + " is out of bounds. Total laps: " + lapStartTimes.size());
        }
        long startTime = lapStartTimes.get(lapIndex);
        long endTime = lapIndex < lapStartTimes.size() - 1
                ? lapStartTimes.get(lapIndex + 1) : System.currentTimeMillis();
        double duration = (endTime - startTime) / 1000.0;
        return UnitToString.secondsToTime(duration);
    }

    public void startLogging() {
        if (running) {
            return;
        }
        running = true;
        for (Metric metric : metrics) {
            metric.startLogging();
        }
        if (lapStartTimes.isEmpty()) {
            lapStartTimes.add(System.currentTimeMillis());
        }
    }

    public void stopLogging() {
        if (!running) {
            return;
        }
        running = false;
        for (Metric metric : metrics) {
            metric.stopLogging();
        }
    }

    public boolean isRunning() {
        return running;
    }

    @Override
    public ILocationService getLocationService() {
        return locationService;
    }

    @Override
    public IHeartRateSensorService getHeartRateSensorService() {
        return heartRateSensorService;
    }

    @Override
    public IPowerMeterService getPowerMeterService() {
        return powerMeterService;
    }

    @Override
    public ISpeedSensorService getSpeedSensorService() {
        return speedSensorService;
    }

    public Metric getByName(String name) {
        for (Metric metric : metrics) {
            if (metric.getName().equals(name)) {
                return metric;
            }
        }
        return null;
    }

    public List<String> getNames() {
        List<String> names = new ArrayList<>();
        for (Metric metric : metrics) {
            names.add(metric.getName());
        }
        return names;
    }

    public Metric getByMetricType(MetricType metricType) {
        for (Metric metric : metrics) {
            if (metric.getType() == metricType) {
                return metric;
            }
        }
        throw new IllegalArgumentException("Metric with type " + metricType + " not found");
    }
}
